storage/layerupload.go

```diff
@@ -337,85 +170,47 @@ func (luc *layerUploadController) linkLayer(digest digest.Digest) error {
 	return luc.layerStore.driver.PutContent(layerLinkPath, []byte(digest))
 }
 
-// localFSLayerUploadStore implements a local layerUploadStore. There are some
-// complexities around hashsums that make round tripping to the storage
-// backend problematic, so we'll store and read locally for now. By GO-beta,
-// this should be fully implemented on top of the backend storagedriver.
-//
-// For now, the directory layout is as follows:
-//
-// 	/<temp dir>/registry-layer-upload/
-// 		<uuid>/
-// 			-> state.json
-// 			-> data
-//
-// Each upload, identified by uuid, has its own directory with a state file
-// and a data file. The state file has a json representation of the current
-// state. The data file is the in-progress upload data.
-type localFSLayerUploadStore struct {
-	root string
-}
-
-func newTemporaryLocalFSLayerUploadStore() (layerUploadStore, error) {
-	path, err := ioutil.TempDir("", "registry-layer-upload")
+// removeResources should clean up all resources associated with the upload
+// instance. An error will be returned if the clean up cannot proceed. If the
+// resources are already not present, no error will be returned.
+func (luc *layerUploadController) removeResources() error {
+	startedAtPath, err := luc.layerStore.pathMapper.path(uploadStartedAtPathSpec{
+		name: luc.name,
+		uuid: luc.uuid,
+	})
 
 	if err != nil {
-		return nil, err
-	}
-
-	return &localFSLayerUploadStore{
-		root: path,
-	}, nil
-}
-
-func (llufs *localFSLayerUploadStore) New(name string) (LayerUploadState, error) {
-	lus := LayerUploadState{
-		Name: name,
-		UUID: uuid.New(),
-	}
-
-	if err := os.Mkdir(llufs.path(lus.UUID, ""), 0755); err != nil {
-		return lus, err
+		return err
 	}
 
-	return lus, nil
-}
-
-func (llufs *localFSLayerUploadStore) Open(uuid string) (layerFile, error) {
-	fp, err := os.OpenFile(llufs.path(uuid, "data"), os.O_CREATE|os.O_APPEND|os.O_RDWR, 0644)
+	dataPath, err := luc.layerStore.pathMapper.path(uploadDataPathSpec{
+		name: luc.name,
+		uuid: luc.uuid,
+	})
 
 	if err != nil {
-		return nil, err
-	}
-
-	return fp, nil
-}
-
-func (llufs *localFSLayerUploadStore) GetState(uuid string) (LayerUploadState, error) {
-	var lus LayerUploadState
-
-	if _, err := os.Stat(llufs.path(uuid, "")); err != nil {
-		if os.IsNotExist(err) {
-			return lus, ErrLayerUploadUnknown
-		}
-
-		return lus, err
+		return err
 	}
-	return lus, nil
-}
 
-func (llufs *localFSLayerUploadStore) DeleteState(uuid string) error {
-	if err := os.RemoveAll(llufs.path(uuid, "")); err != nil {
-		if os.IsNotExist(err) {
-			return ErrLayerUploadUnknown
+	// Some backends may still have a directory entry for the upload. Resolve
+	// and delete that beast! It may not exist, but it's okay.
+	dirPath := path.Dir(dataPath)
+
+	for _, path := range []string{startedAtPath, dataPath, dirPath} {
```

Comment:
You can actually get away with a single `driver.Delete` here if you provide `dirPath` given that it is recursive. Maybe we don't want to assume that `startedAtPath` is a child of `dirPath`, but the delete of `dataPath` is certainly redundant.